### ISSUE TITLE
Unable to authenticate against Openstack with SAML auth configured

### ISSUE THREAD

**Issue by khahmed** [openshift/installer#2769]
# Version

```console
$ openshift-install version
./openshift-install v4.2.8
built from commit 425e4ff0037487e32571258640b39f56d5ee5572
release image quay.io/openshift-release-dev/ocp-release@sha256:4bf307b98beba4d42da3316464013eac120c6e5a398646863ef92b0e2c621230
```

# Platform:

# What happened?

I'm having trouble getting the installer to work with my OpenStack. The OpenStack authentication is configured to use SAML. I'm not sure I have something configured incorrectly so any advice is appreciated if I'm doing something wrong. The openstack CLI works but the openshift-installer doesn't seem. I don't control the OpenStack installation so if there is something that needs to be done on that side  please advise and I can ask the admins. I know the Openstack authentication is configured to use Keycloak. 

~~~
(openstack_cli) root@icp-master:~/openshift/4.2# OS_CLOUD=openstack ./openshift-install --dir=foo create install-config --log-level debug
DEBUG OpenShift Installer v4.2.8
DEBUG Built from commit 425e4ff0037487e32571258640b39f56d5ee5572
DEBUG Fetching "Install Config"...
DEBUG Loading "Install Config"...
DEBUG   Loading "SSH Key"...
DEBUG   Loading "Base Domain"...
DEBUG     Loading "Platform"...
DEBUG   Loading "Cluster Name"...
DEBUG     Loading "Base Domain"...
DEBUG   Loading "Pull Secret"...
DEBUG   Loading "Platform"...
DEBUG   Fetching "SSH Key"...
DEBUG   Generating "SSH Key"...
? SSH Public Key /home/ubuntu/.ssh/id_rsa.pub
DEBUG   Fetching "Base Domain"...
DEBUG     Fetching "Platform"...
DEBUG     Generating "Platform"...
? Platform openstack
? Cloud openstack
FATAL failed to fetch Install Config: failed to fetch dependency of "Install Config": failed to fetch dependency of "Base Domain": failed to generate asset "Platform": You must provide a password to authenticate

(openstack_cli) root@icp-master:~/openshift/4.2# cat clouds.yaml
clouds:
  openstack:
    auth_type: v3samlpassword
    auth:
      auth_url: https://pollux.cscs.ch:13000/v3
      project_id:  XXXXX
      project_name:  "XXX"
      project_domain_name: "cscs"
    username: "XXXXX"
    password: "XXXXX"
    # Federation Identity Provider (IdP) can be found using the following:
    # openstack identity provider list
    identity_provider: cscskc
    # Federation protocol can be found using the following:
    # openstack federation protocol list --identity-provider "__MYIDP__"
    protocol: mapped
    identity_provider_url: https://auth.cscs.ch/auth/realms/cscs/protocol/saml/
    region_name: "Pollux"
    interface: "public"
    version: 3


~~~
# What you expected to happen?

The installer should be able to authenticate and retrieve neccessary information from Openstack.

# How to reproduce it (as minimally and precisely as possible)?

```console
$ your-commands-here
```

# Anything else we need to know?

Enter text here.

# References

- enter text here.

**Comment by Fedosin** [comment]
From the example above I can say that `clouds.yaml` is malformed. Gophercloud looks for a password and a name in the `auth` section and cannot find them there. Here you can find some examples: https://docs.openstack.org/python-openstackclient/train/configuration/index.html#configuration-files

**Comment by khahmed** [comment]
@Fedosin I tried changing the indentation of user/password but got a different error. Do you know what the DomainID or DomainName is referring to? I tried commenting out the project_id but it didn't work. Note that the openstack CLI is able to authenticate and retrieve results.  I was referred to this blog for the format of clouds.yaml for authentication with SAML.

https://joscor.com/blog/use-openstack-cli-with-cloud-configuration-file-saml-federation/

Maybe its not accurate for my environment. Any further suggestions appreciated. 

~~~
(openstack_cli) root@icp-master:~/openshift/4.2# OS_CLOUD=openstack ./openshift-install create install-config --log-level debug
DEBUG OpenShift Installer v4.2.8
DEBUG Built from commit 425e4ff0037487e32571258640b39f56d5ee5572
DEBUG Fetching "Install Config"...
DEBUG Loading "Install Config"...
DEBUG   Loading "SSH Key"...
DEBUG   Loading "Base Domain"...
DEBUG     Loading "Platform"...
DEBUG   Loading "Cluster Name"...
DEBUG     Loading "Base Domain"...
DEBUG   Loading "Pull Secret"...
DEBUG   Loading "Platform"...
DEBUG   Fetching "SSH Key"...
DEBUG   Generating "SSH Key"...
? SSH Public Key /home/ubuntu/.ssh/id_rsa.pub
DEBUG   Fetching "Base Domain"...
DEBUG     Fetching "Platform"...
DEBUG     Generating "Platform"...
? Platform openstack
? Cloud openstack
FATAL failed to fetch Install Config: failed to fetch dependency of "Install Config": failed to fetch dependency of "Base Domain": failed to generate asset "Platform": You must provide exactly one of DomainID or DomainName to authenticate by Username
(openstack_cli) root@icp-master:~/openshift/4.2# cat clouds.yaml
clouds:
  openstack:
    auth_type: v3samlpassword
    #auth_url: https://pollux.cscs.ch:13000/v3
    auth:
      auth_url: https://pollux.cscs.ch:13000/v3
      #project_id: XXXXX
      project_name:  "XXX"
      project_domain_name: "XXX"
      username: "XXXX"
      password: "XXXX"
    region_name: Pollux
    # Federation Identity Provider (IdP) can be found using the following:
    # openstack identity provider list
    identity_provider: cscskc
    # Federation protocol can be found using the following:
    # openstack federation protocol list --identity-provider "__MYIDP__"
    protocol: mapped
    identity_provider_url: https://auth.cscs.ch/auth/realms/cscs/protocol/saml/
    interface: "public"
    version: 3
(openstack_cli) root@icp-master:~/openshift/4.2# OS_CLOUD=openstack openstack region list
+--------+---------------+-------------+
| Region | Parent Region | Description |
+--------+---------------+-------------+
| Pollux | None          |             |
+--------+---------------+-------------+
~~~

**Comment by iamemilio** [comment]
Does your `clouds.yaml` have a domain in it? https://docs.openstack.org/python-openstackclient/pike/configuration/index.html
> If you are using Identity v3 you need to specify the user and the project domain name

**Comment by Fedosin** [comment]
Long story short, I have to admit that [gophercloud](https://github.com/gophercloud/gophercloud), the library we use to communicate with OpenStack, doesn't support any other authentication flow except the default one, i.e. the username-password. We need to open an RFE to add these features there before we can start supporting them in OpenShift. All available config parameters are listed [here](https://github.com/gophercloud/utils/blob/master/openstack/clientconfig/results.go).

**Comment by khahmed** [comment]
@Fedosin I looked through the gophercloud issues and found some issues that might be related:
https://github.com/gophercloud/gophercloud/issues/1005
I think token support is what is needed to enable SAML though I'm not well versed on this.  I tried another test where I set the OS_TOKEN after the SAML login and re-run the install. It seems to get further than it did before. Its able to get the network and flavor information but then fails. In issue 1005 they talk about "The key here is "external packages" - I'd like for applications which consume Gophercloud to start using clientconfig for authentication since that package will support all authentication options users/devs expect to see. " Does the opneshift installer already use clientconfig? 

~~~

(openstack_cli) root@icp-master:~/openshift/4.2# source ~/openstack_cli/openstack/cli/pollux.env
 * Creating environment for openstack CLI:
Username: kahmed
Password:
[openstack --os-auth-type v3samlpassword token issue]
 * Got an unscoped token, preparing environment...
 * Logged in user ID kahmed: 6e565d2b1e56458c827bf546dd3c406f
[openstack project list]
 * Selected project g43:  XXXX
[openstack --os-project-id XXXXXXX  token issue]
 * Setting custom 'swift' alias
 * Setting custom 'cinder' alias
 * Environment ready for openstack CLI with scoped project: g43

(openstack_cli) root@icp-master:~/openshift/4.2# OS_CLOUD=openstack ./openshift-install create install-config --log-level debug
DEBUG OpenShift Installer v4.2.8
DEBUG Built from commit 425e4ff0037487e32571258640b39f56d5ee5572
DEBUG Fetching "Install Config"...
DEBUG Loading "Install Config"...
DEBUG   Loading "SSH Key"...
DEBUG   Loading "Base Domain"...
DEBUG     Loading "Platform"...
DEBUG   Loading "Cluster Name"...
DEBUG     Loading "Base Domain"...
DEBUG   Loading "Pull Secret"...
DEBUG   Loading "Platform"...
DEBUG   Fetching "SSH Key"...
DEBUG   Generating "SSH Key"...
? SSH Public Key /home/ubuntu/.ssh/id_rsa.pub
DEBUG   Fetching "Base Domain"...
DEBUG     Fetching "Platform"...
DEBUG     Generating "Platform"...
? Platform openstack
? Cloud openstack
? ExternalNetwork ext-net1
? APIFloatingIPAddress 148.187.97.168
? FlavorName m1.large
FATAL failed to fetch Install Config: failed to fetch dependency of "Install Config": failed to fetch dependency of "Base Domain": failed to generate asset "Platform": You must provide exactly one of DomainID or DomainName in a Scope with ProjectName
(openstack_cli) root@icp-master:~/openshift/4.2# env | grep OS_
OS_PROJECT_ID=XXXXX
OS_IDENTITY_PROVIDER=cscskc
OS_IDENTITY_PROVIDER_URL=https://auth.cscs.ch/auth/realms/cscs/protocol/saml/
OS_IDENTITY_API_VERSION=3
OS_AUTH_TYPE=token
OS_PROTOCOL=mapped
OS_AUTH_URL=https://pollux.cscs.ch:13000/v3
OS_TOKEN=XXXXX
OS_USERNAME=kahmed
OS_INTERFACE=public
~~~
